lib/statsd/instrument/aggregator.rb

```diff
@@ -0,0 +1,246 @@
+# frozen_string_literal: true
+
+module StatsD
+  module Instrument
+    class Aggregator
+      CONST_SAMPLE_RATE = 1.0
+      COUNT = :c
+      DISTRIBUTION = :d
+      MEASURE = :ms
+      HISTOGRAM = :h
+      GAUGE = :g
+
+      class << self
+        def finalize(aggregation_state, sink, datagram_builders, datagram_builder_class)
+          proc do
+            aggregation_state.each do |_key, counter|
+              if datagram_builders[counter[:no_prefix]].nil?
+                datagram_builders[counter[:no_prefix]] =
+                  create_datagram_builder(datagram_builder_class, no_prefix: counter[:no_prefix])
+              end
+              sink << datagram_builders[counter[:no_prefix]].c(
+                counter[:name],
+                counter[:value],
+                CONST_SAMPLE_RATE,
+                counter[:tags],
+              )
+            end
+            aggregation_state.clear
+          end
+        end
+
+        private
+
+        def create_datagram_builder(builder_class, no_prefix:)
+          builder_class.new(
+            prefix: no_prefix ? nil : @metric_prefix,
+            default_tags: @default_tags,
+          )
+        end
+      end
+
+      # @param sink [#<<] The sink to write the aggregated metrics to.
+      # @param datagram_builder_class [Class] The class to use for building datagrams.
+      # @param prefix [String] The prefix to add to all metrics.
+      # @param default_tags [Array<String>] The tags to add to all metrics.
+      # @param flush_interval [Float] The interval at which to flush the aggregated metrics.
+      # @param max_values [Integer] The maximum number of values to aggregate before flushing.
+      def initialize(sink, datagram_builder_class, prefix, default_tags, flush_interval: 5.0, max_values: 100)
+        @sink = sink
+        @datagram_builder_class = datagram_builder_class
+        @metric_prefix = prefix
+        @default_tags = default_tags
+        @datagram_builders = {
+          true: nil,
+          false: nil,
+        }
+        @max_values = max_values
+
+        # Mutex protects the aggregation_state hash from concurrent access
+        @mutex = Mutex.new
+        @aggregation_state = {}
+
+        @pid = Process.pid
+        @flush_interval = flush_interval
+        @flush_thread = Thread.new do
+          Thread.current.abort_on_exception = true
+          loop do
+            sleep(@flush_interval)
+            thread_healthcheck
+            flush
+          end
+        end
+
+        ObjectSpace.define_finalizer(
+          self,
+          self.class.finalize(@aggregation_state, @sink, @datagram_builders, @datagram_builder_class),
+        )
```

Comment:
I'd heavily advise against using a finalizer, especially one that does IOs.

Reply:
`BatchedDUPSink` has one, but honestly I think it was a mistake.

Reply:
I see, I got the idea from our Sink code, but do you know of a better option in this case? I am not so versed with this part of Ruby if I am being honest.

Reply:
I presume the better option would just be to let the aggregated lines drop on the floor, and not bother trying to export them. In the common case of a `fork` we can make sure that we `flush` appropriately, and otherwise just let the data go away.

I'd be interested to know more about why finalizers are not something you'd recommend as well; it's hard to find good references about this stuff. ❤️

Reply:
Because finalizers can't be run concurrently, so doing potentially long things in a finalizer may stall execution for a long time.

The reason I did it in `UDPBatchSink` was to not lose events on exit, but I think we need to start accepting that StatsD events are best effort.

Reply:
After reading the comment here, I went on to search and [read here](https://blog.ycrash.io/problems-with-finalizer/) that it can slow down GC activity as well.